### PR TITLE
objstore: Added WithExpectedErrs which allows to control instrumentation (e.g not increment failures for expected not found).

### PR DESCRIPTION
**Pull request by bwplotka**
This 3rd alternative to below PRs that was discussed offline: https://cloud-native.slack.com/archives/CL25937SP/p1585903230146500

Closes https://github.com/thanos-io/thanos/pull/2365 and https://github.com/thanos-io/thanos/pull/2369

This allows to not wake up on-call in the middle of the night, because of the expected, properly handled case (:

Thanks @pracucci and @pstibrany for starting this work and discussion. We were aware of this for like 12 months, but only with your discussions, we clarified what is expected here :hugs:  Your input was super valuable.

## Changes

Added new interface:

```go
type InstrumentedBucket interface {
	Bucket

	// WithExpectedErrs allows to specify a filter that marks certain errors as expected, so it will not wake up
	// on-call engineer (:
	WithExpectedErrs(IsOpFailureExpectedFunc) Bucket
```

Also: Has to move inmem to objstore for testing.

Signed-off-by: Bartlomiej Plotka <bwplotka@gmail.com>

### CODE REVIEW

File: pkg/objstore/azure/azure.go [hunk 1]
```diff
@@ -64,7 +64,7 @@ func (conf *Config) validate() error {
 	return nil
 }
 
-// NewBucket returns a new Bucket using the provided Azure config.
+// NewInMemBucket returns a new Bucket using the provided Azure config.
```
Comment (GiedriusS) [inline pkg/objstore/azure/azure.go:67]: Seems like an unrelated change?

File: pkg/objstore/client/factory.go [hunk 2]
```diff
@@ -40,9 +40,9 @@ type BucketConfig struct {
 	Config interface{} `yaml:"config"`
 }
 
-// NewBucket initializes and returns new object storage clients.
+// NewInMemBucket initializes and returns new object storage clients.
```
Comment (GiedriusS) [inline pkg/objstore/client/factory.go:43]: Seems like an unrelated change?

File: pkg/objstore/filesystem/filesystem.go [hunk 3]
```diff
@@ -43,7 +43,7 @@ func NewBucketFromConfig(conf []byte) (*Bucket, error) {
 	return NewBucket(c.Directory)
 }
 
-// NewBucket returns a new filesystem.Bucket.
+// NewInMemBucket returns a new filesystem.Bucket.
```
Comment (GiedriusS) [inline pkg/objstore/filesystem/filesystem.go:46]: Ditto.

File: pkg/objstore/gcs/gcs.go [hunk 4]
```diff
@@ -41,7 +41,7 @@ type Bucket struct {
 	closer io.Closer
 }
 
-// NewBucket returns a new Bucket against the given bucket handle.
+// NewInMemBucket returns a new Bucket against the given bucket handle.
```
Comment (GiedriusS) [inline pkg/objstore/gcs/gcs.go:44]: Ditto.

File: pkg/objstore/oss/oss.go [hunk 5]
```diff
@@ -150,7 +150,7 @@ func (b *Bucket) ObjectSize(ctx context.Context, name string) (uint64, error) {
 	return 0, errors.New("content-length header not found")
 }
 
-// NewBucket returns a new Bucket using the provided oss config values.
+// NewInMemBucket returns a new Bucket using the provided oss config values.
```
Comment (GiedriusS) [inline pkg/objstore/oss/oss.go:153]: Ditto.

File: pkg/objstore/azure/azure.go (hunk 1 above)
Comment (bwplotka) [reply pkg/objstore/azure/azure.go:67]: :man_facepalming:

File: pkg/block/fetcher.go [hunk 6]
```diff
@@ -236,7 +236,7 @@ func (f *BaseFetcher) loadMeta(ctx context.Context, id ulid.ULID) (*metadata.Met
 		}
 	}
 
-	r, err := f.bkt.Get(ctx, metaFile)
+	r, err := f.bkt.ReaderWithExpectedErrs(f.bkt.IsObjNotFoundErr).Get(ctx, metaFile)
```
Comment (pracucci) [inline pkg/block/fetcher.go:239]: Is this expected to cover the case of a partially uploaded block (while still uploading)?

Comment (bwplotka) [reply pkg/block/fetcher.go:239]: Yes

**Comment by bwplotka** [comment]
Exactly this + Giedrius addressed comments was merged in release 0.12 and will be ported to master soon: https://github.com/thanos-io/thanos/commit/c05daeeace272a94fae81fd0c15ff70adab50e10
